### ISSUE TITLE
Invalid session state (not "secure" related)

### ISSUE THREAD

**Issue by michal-rybinski** [vouch/vouch-proxy#115]
Hi all,

I've set up a vouch in front of my services and are using AWS Cognito as an oath2 provider.
I managed to get it working with Benjamin's help and all works as expected, but from time to time I am encountering an "Invalid session state" issue which by the looks of it points to either unicode encode/decode problem or something else that is dealing with extracting data from cookies on the vouch side. This is the log entry that appears when the problem surfaces and it is quite clear that the problem is with comparison of session variable stored locally and received from cookie:

`{"level":"error","ts":1556534349.1250064,"msg":"Invalid session state: stored NHEDnfUelEV/FUfe4gajRJUptDzdG3cH/+tvaBp0YNs=, returned NHEDnfUelEV/FUfe4gajRJUptDzdG3cH/ tvaBp0YNs="}`

every time there is a "+" sign generated in session state variable, it is being converted to a space or deleted from returned session state variable what cases the evaluation to fail and presentation of the mentioned above error.

Has someone else encountered this problem as well?

Thanks,
Michal

**Comment by michal-rybinski** [comment]
Immediately after the above message there is another one which shows that the plus sign have been provided in the referer string as properly urlencoded string:

`{"level":"info","ts":1556534349.125064,"msg":"|200|  205.705µs /auth","statusCode":200,"request":170,"latency":0.000205705,"avgLatency":0.003134225,"ipPort":"127.0.0.1:34346","method":"GET","host":"login.xxx.com","path":"/auth","referer":"https://auth.xxx.com/login?client_id=xxx&redirect_uri=https%3A%2F%2Flogin.xxx.com%2Fauth&response_type=code&scope=openid+email+profile&state=NHEDnfUelEV%2FFUfe4gajRJUptDzdG3cH%2F%2BtvaBp0YNs%3D"}`

What can be easily checked here: https://www.urldecoder.org/

**Comment by bnfinet** [comment]
@michal-rybinski could you please upgrade to `v0.5.13` and (hopefully) confirm the fix

**Comment by michal-rybinski** [comment]
Unfortunately it is still there...  

`$ sudo ./vouch-proxy -loglevel debug
{"level":"debug","ts":1556877791.5207422,"msg":"logLevel set to debug"}
{"level":"info","ts":1556877791.5208442,"msg":"jwt.secret read from config/secret"}
{"level":"warn","ts":1556877791.5209095,"msg":"generating random session.key"}
{"level":"info","ts":1556877791.521017,"msg":"configuring oidc OAuth with Endpoint https://auth.xxx.com/oauth2/authorize"}
{"level":"debug","ts":1556877791.521037,"msg":"vouch.jwt.secret is 44 characters long"}
{"level":"debug","ts":1556877791.5210414,"msg":"vouch.session.key is 44 characters long"}
{"level":"debug","ts":1556877791.5210474,"msg":"checking availability of tcp port: 0.0.0.0:9090"}
{"level":"debug","ts":1556877791.5216308,"msg":"viper settings map[vouch:map[port:9090 listen:0.0.0.0 allowallusers:true cookie:map[name:VouchCookie httponly:true domain:xxx.com] db:map[file:data/vouch_bolt.db] loglevel:info session:map[name:VouchSession] headers:map[querystring:access_token redirect:X-Vouch-Requested-URI jwt:X-Vouch-Token] jwt:map[compress:true maxage:240 issuer:Vouch]] oauth:map[user_info_url:https://auth.xxx.com/oauth2/userInfo provider:oidc client_id:xxx callback_url:https://login.xxx.com/auth client_secret:xxx auth_url:https://auth.xxx.com/oauth2/authorize token_url:https://auth.xxx.com/oauth2/token scopes:[openid email profile]]]"}
{"level":"debug","ts":1556877791.5216935,"msg":"opening dbfile data/vouch_bolt.db"}
{"level":"info","ts":1556877791.5225136,"msg":"starting Vouch","version":"undefined","buildtime":"undefined","buildhost":"undefined","branch":"undefined","semver":"undefined","listen":"0.0.0.0:9090","oauth.provider":"oidc"}
{"level":"debug","ts":1556877791.5226753,"msg":"serving static files from /static"}
{"level":"debug","ts":1556877795.2861793,"msg":"Request received : &{GET /validate HTTP/1.0 1 0 map[X-Forwarded-Proto:[https] Upgrade-Insecure-Requests:[1] Accept-Encoding:[gzip, deflate] Dnt:[1] Connection:[close] X-Real-Ip:[10.123.68.207] X-Forwarded-For:[10.123.68.207] User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0] Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8] Accept-Language:[en-GB,en;q=0.5]] {} <nil> 0 [] true login.xxx.com map[] map[] <nil> map[] 127.0.0.1:47194 /validate <nil> <nil> <nil> 0xc000176360}"}
{"level":"debug","ts":1556877795.2862089,"msg":"/validate"}
{"level":"error","ts":1556877795.2862206,"msg":"no jwt found in request"}
{"level":"warn","ts":1556877795.2862265,"msg":"domain login.xxx.com not found in any domains []"}
{"level":"debug","ts":1556877795.2862315,"msg":"setting the cookie domain to xxx.com"}
{"level":"debug","ts":1556877795.2862487,"msg":"CaptureWriter.Write set w.StatusCode Ƒ"}
{"level":"debug","ts":1556877795.2862647,"msg":"Request handled successfully: 401"}
{"level":"info","ts":1556877795.2862735,"msg":"|401|   57.437µs /validate","statusCode":401,"request":1,"latency":0.000057437,"avgLatency":0.000057437,"ipPort":"127.0.0.1:47194","method":"GET","host":"login.xxx.com","path":"/validate","referer":""}
{"level":"debug","ts":1556877795.3435104,"msg":"Request received : &{GET /login?url=http://elasticsearch.xxx.com/&vouch-failcount=&X-Vouch-Token=&error= HTTP/1.0 1 0 map[User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0] Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8] Accept-Language:[en-GB,en;q=0.5] Dnt:[1] Connection:[close] Accept-Encoding:[gzip, deflate, br] Upgrade-Insecure-Requests:[1] X-Forwarded-Proto:[https]] {} <nil> 0 [] true login.xxx.com map[] map[] <nil> map[] 127.0.0.1:47196 /login?url=http://elasticsearch.xxx.com/&vouch-failcount=&X-Vouch-Token=&error= <nil> <nil> <nil> 0xc000176570}"}
{"level":"debug","ts":1556877795.343548,"msg":"/login"}
{"level":"warn","ts":1556877795.343555,"msg":"domain login.xxx.com not found in any domains []"}
{"level":"debug","ts":1556877795.3435605,"msg":"setting the cookie domain to xxx.com"}
{"level":"debug","ts":1556877795.3435974,"msg":"session state set to jqVE//CZNg/wo5MUCnJRD3Qg4/cE/CiHbXhKw1nAt+0="}
{"level":"debug","ts":1556877795.343608,"msg":"session requestedURL set to http://elasticsearch.xxx.com/"}
{"level":"debug","ts":1556877795.3436124,"msg":"saving session"}
{"level":"debug","ts":1556877795.3438785,"msg":"redirecting to oauthURL https://auth.xxx.com/oauth2/authorize?client_id=xxx&redirect_uri=https%3A%2F%2Flogin.xxx.com%2Fauth&response_type=code&scope=openid+email+profile&state=jqVE%2F%2FCZNg%2Fwo5MUCnJRD3Qg4%2FcE%2FCiHbXhKw1nAt%2B0%3D"}
{"level":"debug","ts":1556877795.3438952,"msg":"CaptureWriter.Write set w.StatusCode Į"}
{"level":"debug","ts":1556877795.3439088,"msg":"Request handled successfully: 302"}
{"level":"info","ts":1556877795.3439174,"msg":"|302|  367.555µs /login","statusCode":302,"request":2,"latency":0.000367555,"avgLatency":0.000212496,"ipPort":"127.0.0.1:47196","method":"GET","host":"login.xxx.com","path":"/login","referer":""}
{"level":"debug","ts":1556877811.1748483,"msg":"Request received : &{GET /auth?code=747cd80e-5a13-45d3-8c24-66adce5fe082&state=jqVE//CZNg/wo5MUCnJRD3Qg4/cE/CiHbXhKw1nAt+0= HTTP/1.0 1 0 map[Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8] Accept-Language:[en-GB,en;q=0.5] Referer:[https://auth.xxx.com/] Dnt:[1] Cookie:[VouchSession=MTU1Njg3Nzc5NXxEdi1CQkFFQ180SUFBUkFCRUFBQV85RF9nZ0FEQm5OMGNtbHVad3dIQUFWemRHRjBaUVp6ZEhKcGJtY01MZ0FzYW5GV1JTOHZRMXBPWnk5M2J6Vk5WVU51U2xKRU0xRm5OQzlqUlM5RGFVaGlXR2hMZHpGdVFYUXJNRDBHYzNSeWFXNW5EQTRBREhKbGNYVmxjM1JsWkZWU1RBWnpkSEpwYm1jTUtnQW9hSFIwY0RvdkwyVnNZWE4wYVdOelpXRnlZMmd1YjJabWFXTmxMbkp2Ylc5dVpYUXVZMjl0THdaemRISnBibWNNS2dBb2FIUjBjRG92TDJWc1lYTjBhV056WldGeVkyZ3ViMlptYVdObExuSnZiVzl1WlhRdVkyOXRMd05wYm5RRUFnQUN8B6CPnqpbgZQPkfVYGwBsIf6OWB9RvHFl8grol7Ms5oU=] User-Agent:[Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0] X-Forwarded-Proto:[https] Connection:[close] Accept-Encoding:[gzip, deflate, br] Upgrade-Insecure-Requests:[1]] {} <nil> 0 [] true login.xxx.com map[] map[] <nil> map[] 127.0.0.1:47200 /auth?code=747cd80e-5a13-45d3-8c24-66adce5fe082&state=jqVE//CZNg/wo5MUCnJRD3Qg4/cE/CiHbXhKw1nAt+0= <nil> <nil> <nil> 0xc000176c90}"}
{"level":"debug","ts":1556877811.1748862,"msg":"/auth"}
{"level":"error","ts":1556877811.1750772,"msg":"Invalid session state: stored jqVE//CZNg/wo5MUCnJRD3Qg4/cE/CiHbXhKw1nAt+0=, returned jqVE//CZNg/wo5MUCnJRD3Qg4/cE/CiHbXhKw1nAt 0="}
{"level":"debug","ts":1556877811.175179,"msg":"CaptureWriter.Write set w.StatusCode È"}
{"level":"debug","ts":1556877811.1752214,"msg":"Request handled successfully: 200"}
{"level":"info","ts":1556877811.175229,"msg":"|200|  336.045µs /auth","statusCode":200,"request":3,"latency":0.000336045,"avgLatency":0.000253679,"ipPort":"127.0.0.1:47200","method":"GET","host":"login.xxx.com","path":"/auth","referer":"https://auth.xxx.com/"}`

**Comment by michal-rybinski** [comment]
and I did pull and build the newest version:

`# git pull
remote: Enumerating objects: 74, done.
remote: Counting objects: 100% (74/74), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 87 (delta 46), reused 68 (delta 41), pack-reused 13
Unpacking objects: 100% (87/87), done.
From https://github.com/vouch/vouch-proxy
   998906f..229f735  master     -> origin/master
 * [new tag]         v0.5.15    -> v0.5.15
From https://github.com/vouch/vouch-proxy
 * [new tag]         v0.5.13    -> v0.5.13
 * [new tag]         v0.5.14    -> v0.5.14
Updating 998906f..229f735
Fast-forward
 README.md                         |  13 +++++++++++--
 config/config.yml_example         |  13 +++++++++++++
 handlers/handlers.go              | 129 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------
 pkg/cfg/cfg.go                    |  29 +++++++++++++++++++++--------
 pkg/jwtmanager/jwtmanager.go      |  11 ++++++++---
 pkg/jwtmanager/jwtmanager_test.go |  19 +++++++++++++++----
 pkg/structs/structs.go            |   5 +++++
 7 files changed, 181 insertions(+), 38 deletions(-)
# go get ./...
go get: no install location for directory /1/vouch-proxy outside GOPATH
        For more details see: 'go help gopath'
# go build
#`

**Comment by bnfinet** [comment]
`go get: no install location for directory /1/vouch-proxy outside GOPATH`

What's at `$GOPATH/src/github.com/vouch/vouch-proxy`

Your build is looking for packages at `$GOPATH/src/github.com/vouch/vouch-proxy` which I suspect are the old packages.

Please do not push logs and config to gh issues. Please use hastebin as per these instructions from the README...

https://github.com/vouch/vouch-proxy#okay-i-looked-at-the-issues-and-have-tried-some-things-with-my-configs-but-i-still-cant-figure-it-out

**Comment by michal-rybinski** [comment]
sorry about not using hastebin, I've fixed the post now.
As for the issue, you are right, I wasn't aware that source needs to be in a certain place for go to pick it up and I was running it from a completely different directory. Once I've copied it across to the `$GOPATH/src/github.com/vouch/vouch-proxy` I was able to build it without this error what I assume is a good sign :)

I've done some more testing and it looks like it is fixed, as I haven't noticed any "+" signs in the generated sesion state variables so subsequently no problems with decode/encode on them.

Thanks for your help!
